In [4]:
import pandas as pd
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB  
from sklearn.metrics import accuracy_score    
from sklearn.metrics import confusion_matrix 
from sklearn.preprocessing import LabelEncoder

In [17]:
data=pd.read_csv('train.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [18]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 889 entries, 0 to 888
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  889 non-null    int64  
 1   Survived     889 non-null    int64  
 2   Pclass       889 non-null    int64  
 3   Name         889 non-null    object 
 4   Sex          889 non-null    object 
 5   Age          889 non-null    float64
 6   SibSp        889 non-null    int64  
 7   Parch        889 non-null    int64  
 8   Ticket       889 non-null    object 
 9   Fare         889 non-null    float64
 10  Cabin        202 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.5+ KB


### Changing categorical

In [19]:
encoder=LabelEncoder()
data['Sex']=encoder.fit_transform(data['Sex'])
data['Embarked']=encoder.fit_transform(data['Embarked'])
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name  Sex   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris    1  22.0      1      0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...    0  38.0      1      0   
2                             Heikkinen, Miss. Laina    0  26.0      0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  35.0      1      0   
4                           Allen, Mr. William Henry    1  35.0      0      0   

             Ticket     Fare Cabin  Embarked  
0         A/5 21171   7.2500   NaN         2  
1          PC 17599  71.2833   C85         0  
2  STON/O2. 3101282   7.9250   NaN         2  
3            113803  53.1000  C123         2  
4            373450   8.0500   NaN         2

### Checking for Null values and dropping irrelevant features

In [20]:
data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [21]:
data=data.drop(['Cabin','Name','PassengerId','Ticket'],axis=1)
data.head()

Survived  Pclass  Sex   Age  SibSp  Parch     Fare  Embarked
0         0       3    1  22.0      1      0   7.2500         2
1         1       1    0  38.0      1      0  71.2833         0
2         1       3    0  26.0      0      0   7.9250         2
3         1       1    0  35.0      1      0  53.1000         2
4         0       3    1  35.0      0      0   8.0500         2

### Writing one common function

In [29]:
def naive(x,y):
    x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)
    clf=GaussianNB()
    y_pred=clf.fit(x_train,y_train).predict(x_test)
    print('Accuracy Score: ',accuracy_score(y_test,y_pred,normalize=True))
    print('Confusion Matrix: \n',confusion_matrix(y_test,y_pred))

### Survived VS Remaining IDV

In [30]:
y=data['Survived']
X=data[['Pclass','Sex','Age','SibSp',
       'Parch','Fare','Embarked']]

In [31]:
print(naive(X,y))

Accuracy Score:  0.7528089887640449
Confusion Matrix: 
 [[121  36]
 [ 30  80]]
None


### Conclusion

__out of 267 records 201 are classified successfully, model accuracy is 75.2%__

### Pclass vs Remaining IDV

In [32]:
y=data['Pclass']
X=data[['Survived','Sex','Age','SibSp',
       'Parch','Fare','Embarked']]

In [33]:
print(naive(X,y))

Accuracy Score:  0.6966292134831461
Confusion Matrix: 
 [[ 52  13   5]
 [  5  25  19]
 [  5  34 109]]
None


### Conclusion

__out of 267 records 186 are classified successfully, model accuracy is 69.66%__

### Gender vs Remaining IDV

In [35]:
y=data['Sex']
X=data[['Survived','Pclass','Age','SibSp',
       'Parch','Fare','Embarked']]

In [36]:
print(naive(X,y))

Accuracy Score:  0.7153558052434457
Confusion Matrix: 
 [[ 52  46]
 [ 30 139]]
None


### Conclusion
__out of 267 records 191 are classified successfully, model accuracy is 71.5%__

### SibSp vs Remaining IDV

In [37]:
y=data['SibSp']
X=data[['Survived','Pclass','Age','Sex',
       'Parch','Fare','Embarked']]

In [38]:
print(naive(X,y))

Accuracy Score:  0.5543071161048689
Confusion Matrix: 
 [[130   8   2  40   2   0   0]
 [ 29   9   0  25   2   0   0]
 [  6   0   1   0   1   0   0]
 [  1   0   0   3   3   0   0]
 [  0   0   0   0   2   0   0]
 [  0   0   0   0   0   1   0]
 [  0   0   0   0   0   0   2]]
None


### Conclusion
__out of 267 records 148 are classified successfully, model accuracy is 55.4%__

### Parch vs Remaining IDV

In [39]:
y=data['Parch']
X=data[['Survived','Pclass','Age','Sex',
       'SibSp','Fare','Embarked']]

In [40]:
print(naive(X,y))

Accuracy Score:  0.5805243445692884
Confusion Matrix: 
 [[143   9   2  31  13]
 [ 17   5   2  12   4]
 [  7   5   7   6   2]
 [  0   0   0   0   1]
 [  0   0   0   1   0]]
None


### Conclusion
__out of 267 records 155 are classified successfully, model accuracy is 58.05%__

### Embarked vs Remaining IDV

In [41]:
y=data['Embarked']
X=data[['Survived','Pclass','Age','Sex',
       'SibSp','Fare','Parch']]

In [42]:
print(naive(X,y))

Accuracy Score:  0.6928838951310862
Confusion Matrix: 
 [[ 23   3  23]
 [  0  10  12]
 [ 23  21 152]]
None


### Conclusion

__out of 267 records 185 are classified successfully, model accuracy is 69.28%__